File: scripts/audit_lessons.py
"""
audit_lessons.py — one-shot audit, no code modifications.

Checks whether LESSONS_META[i].dialogue can be losslessly derived from
LESSONS[i].sentences via contextAbove / contextBelow. Outputs a per-lesson
report so we know which lessons are safe to migrate to a single source.

Usage: python3 scripts/audit_lessons.py
"""
import json5
import pathlib
import re
import sys

HTML = pathlib.Path(__file__).parent.parent / "public" / "index.html"
src  = HTML.read_text(encoding="utf-8")


def extract_array_literal(text, decl_name):
    """Find `const NAME = [...]` and return the array literal string."""
    marker = f"const {decl_name} = ["
    start  = text.find(marker)
    if start == -1:
        raise RuntimeError(f"not found: {decl_name}")
    i = start + len(marker) - 1   # position of opening '['
    depth = 0
    in_str = None       # None | "'" | '"' | '`'
    in_line_cmt  = False
    in_block_cmt = False
    while i < len(text):
        ch  = text[i]
        nxt = text[i + 1] if i + 1 < len(text) else ""
        if in_line_cmt:
            if ch == "\n": in_line_cmt = False
            i += 1; continue
        if in_block_cmt:
            if ch == "*" and nxt == "/":
                in_block_cmt = False; i += 2; continue
            i += 1; continue
        if in_str:
            if ch == "\\": i += 2; continue
            if ch == in_str: in_str = None
            i += 1; continue
        if ch == "/" and nxt == "/": in_line_cmt  = True; i += 2; continue
        if ch == "/" and nxt == "*": in_block_cmt = True; i += 2; continue
        if ch in ("'", '"', "`"): in_str = ch; i += 1; continue
        if ch == "[":
            depth += 1
        elif ch == "]":
            depth -= 1
            if depth == 0:
                return text[start + len(marker) - 1 : i + 1]
        i += 1
    raise RuntimeError(f"unterminated array: {decl_name}")


LESSONS_RAW = extract_array_literal(src, "LESSONS")
META_RAW    = extract_array_literal(src, "LESSONS_META")

# json5 handles: unquoted keys, single quotes, trailing commas, comments.
# But it doesn't handle JS template literals — confirm none present.
for name, raw in (("LESSONS", LESSONS_RAW), ("LESSONS_META", META_RAW)):
    if "`" in raw:
        print(f"warn: backticks in {name}; json5 may fail", file=sys.stderr)

try:
    LESSONS      = json5.loads(LESSONS_RAW)
    LESSONS_META = json5.loads(META_RAW)
except Exception as e:
    print("json5 parse error:", e)
    sys.exit(1)

print(f"Loaded {len(LESSONS)} LESSONS and {len(LESSONS_META)} LESSONS_META.\n")


# ── derive dialogue from sentences ────────────────────────────────────
def derive_dialogue(sentences):
    out = []
    last_local_zh = None
    for s in sentences:
        ca = s.get("contextAbove")
        if ca and ca.get("zh") != last_local_zh:
            out.append({
                "role": "local",
                "speaker": ca.get("speaker", ""),
                "zh": ca.get("zh", ""),
            })
            last_local_zh = ca.get("zh")
        out.append({
            "role": s.get("role", "learner"),
            "speaker": s.get("speaker", ""),
            "zh": s.get("zh", ""),
        })
        cb = s.get("contextBelow")
        if cb:
            out.append({
                "role": "local",
                "speaker": cb.get("speaker", ""),
                "zh": cb.get("zh", ""),
            })
            last_local_zh = cb.get("zh")
    return out


def norm_zh(s):
    return re.sub(r"\s+", "", re.sub(r"<br\s*/?>", "", s or "", flags=re.I))


def compare(derived, actual):
    issues = []
    n = max(len(derived), len(actual))
    for i in range(n):
        d = derived[i] if i < len(derived) else None
        a = actual[i]  if i < len(actual)  else None
        if d is None:
            issues.append(f"[{i}] missing in derived (actual: {a.get('speaker','')} | {a.get('zh','')})"); continue
        if a is None:
            issues.append(f"[{i}] missing in actual  (derived: {d.get('speaker','')} | {d.get('zh','')})"); continue
        if norm_zh(d["zh"]) != norm_zh(a.get("zh", "")):
            issues.append(f"[{i}] zh mismatch:\n      derived: {d['zh']}\n      actual:  {a.get('zh','')}")
        if d["role"] != a.get("role"):
            issues.append(f"[{i}] role mismatch: derived={d['role']} actual={a.get('role')}")
    return issues


# ── per-lesson report ─────────────────────────────────────────────────
report = {"perfect": [], "minor": [], "blocked": []}

n = max(len(LESSONS), len(LESSONS_META))
for i in range(n):
    lesson = LESSONS[i] if i < len(LESSONS) else None
    meta   = LESSONS_META[i] if i < len(LESSONS_META) else None
    lid    = lesson and lesson.get("id")
    mid    = meta   and meta.get("id")
    print(f"═══ Lesson {i+1} (LESSONS.id={lid} | META.id={mid}) ═══")

    if not lesson or not meta:
        print("  ⚠️  one side missing\n")
        report["blocked"].append((i, "one side missing")); continue
    sents = lesson.get("sentences") or []
    dlg   = meta.get("dialogue") or []
    if not sents:
        print("  ⚠️  no sentences[]\n"); report["blocked"].append((i, "no sentences")); continue
    if not dlg:
        print("  ⚠️  no dialogue[]\n");  report["blocked"].append((i, "no dialogue"));  continue

    derived = derive_dialogue(sents)
    issues  = compare(derived, dlg)
    print(f"  sentences={len(sents)}  derived={len(derived)}  actual={len(dlg)}")

    if not issues:
        print("  ✅ perfect match\n")
        report["perfect"].append(i)
    elif len(issues) <= 2 and len(derived) == len(dlg):
        print("  🟡 minor diffs:")
        for s in issues: print("    " + s)
        print()
        report["minor"].append((i, issues))
    else:
        print("  ❌ structural mismatch:")
        for s in issues[:8]: print("    " + s)
        if len(issues) > 8: print(f"    ...({len(issues) - 8} more)")
        print()
        report["blocked"].append((i, f"structural ({len(issues)} issues)"))

# ── summary ───────────────────────────────────────────────────────────
print("\n══════════════ SUMMARY ══════════════")
print(f"✅ Perfect (safe to derive): {len(report['perfect'])}  →  lessons {', '.join(str(i+1) for i in report['perfect']) or '(none)'}")
print(f"🟡 Minor diffs (fixable):    {len(report['minor'])}  →  lessons {', '.join(str(o[0]+1) for o in report['minor']) or '(none)'}")
print(f"❌ Blocked (manual review):  {len(report['blocked'])}  →  lessons {', '.join(str(o[0]+1) for o in report['blocked']) or '(none)'}")
print()
if not report["blocked"] and not report["minor"]:
    print("  🟢 ALL lessons can be derived. Safe to migrate.")
elif not report["blocked"]:
    print("  🟡 Most lessons OK; minor diffs can be normalized.")
else:
    print("  🔴 Some lessons need data fixes OR must keep dual structure.")

File: scripts/audit_lessons.js
/**
 * audit_lessons.js — one-shot audit, no code modifications.
 *
 * Goal: for each lesson, check whether LESSONS_META[i].dialogue can be
 * losslessly derived from LESSONS[i].sentences (using contextAbove /
 * contextBelow). Output a per-lesson diff report so we know which lessons
 * are safe to migrate to a single source of truth.
 *
 * Usage: node scripts/audit_lessons.js
 */

const fs = require('fs');
const path = require('path');

const HTML_PATH = path.join(__dirname, '..', 'public', 'index.html');
const src = fs.readFileSync(HTML_PATH, 'utf8');

// ── Step 1. Extract array literals via brace balancing ────────────────
function extractArray(src, declName) {
  const startMarker = 'const ' + declName + ' = [';
  const startIdx = src.indexOf(startMarker);
  if (startIdx === -1) throw new Error('not found: ' + declName);
  let i = startIdx + startMarker.length - 1; // position of opening [
  let depth = 0;
  let inStr = null; // null | "'" | '"' | '`'
  let inLineComment = false;
  let inBlockComment = false;
  for (; i < src.length; i++) {
    const ch = src[i];
    const next = src[i + 1];
    if (inLineComment) { if (ch === '\n') inLineComment = false; continue; }
    if (inBlockComment) { if (ch === '*' && next === '/') { inBlockComment = false; i++; } continue; }
    if (inStr) {
      if (ch === '\\') { i++; continue; }
      if (ch === inStr) inStr = null;
      continue;
    }
    if (ch === '/' && next === '/') { inLineComment = true; i++; continue; }
    if (ch === '/' && next === '*') { inBlockComment = true; i++; continue; }
    if (ch === "'" || ch === '"' || ch === '`') { inStr = ch; continue; }
    if (ch === '[') depth++;
    else if (ch === ']') {
      depth--;
      if (depth === 0) {
        const literal = src.slice(startIdx + ('const ' + declName + ' = ').length, i + 1);
        return literal;
      }
    }
  }
  throw new Error('unterminated array: ' + declName);
}

// Eval JS array literals in a sandbox-ish way. The literals use plain JS
// (no functions or refs), so direct Function eval is safe enough here.
function evalArray(literal) {
  // eslint-disable-next-line no-new-func
  return new Function('return ' + literal + ';')();
}

const LESSONS_RAW = extractArray(src, 'LESSONS');
const META_RAW    = extractArray(src, 'LESSONS_META');
const LESSONS      = evalArray(LESSONS_RAW);
const LESSONS_META = evalArray(META_RAW);

console.log('Loaded', LESSONS.length, 'LESSONS and', LESSONS_META.length, 'LESSONS_META.\n');

// ── Step 2. Derive dialogue from sentences ────────────────────────────
// Heuristic: each sentence carries a learner line (s.zh) and optionally
// contextAbove (local line that precedes it) and contextBelow (local line
// that follows it). Walking sentences in order and emitting:
//   contextAbove(if new) -> learner zh -> contextBelow(if last-or-changes)
// should reproduce a full dialogue.
function deriveDialogue(sentences) {
  const out = [];
  let lastLocalZh = null;
  for (let i = 0; i < sentences.length; i++) {
    const s = sentences[i];
    if (s.contextAbove && s.contextAbove.zh !== lastLocalZh) {
      out.push({
        role: 'local',
        speaker: s.contextAbove.speaker || '',
        zh: s.contextAbove.zh || '',
      });
      lastLocalZh = s.contextAbove.zh;
    }
    out.push({
      role: s.role || 'learner',
      speaker: s.speaker || '',
      zh: s.zh || '',
    });
    if (s.contextBelow) {
      out.push({
        role: 'local',
        speaker: s.contextBelow.speaker || '',
        zh: s.contextBelow.zh || '',
      });
      lastLocalZh = s.contextBelow.zh;
    }
  }
  return out;
}

// ── Step 3. Compare derived vs actual META.dialogue ───────────────────
function normalizeZh(s) {
  return (s || '').replace(/\s+/g, '').replace(/<br\s*\/?>/gi, '');
}

function compareDialogue(derived, actual) {
  const issues = [];
  const n = Math.max(derived.length, actual.length);
  for (let i = 0; i < n; i++) {
    const d = derived[i];
    const a = actual[i];
    if (!d) { issues.push(`[${i}] missing in derived (actual has: ${a.speaker} | ${a.zh})`); continue; }
    if (!a) { issues.push(`[${i}] missing in actual (derived has: ${d.speaker} | ${d.zh})`); continue; }
    if (normalizeZh(d.zh) !== normalizeZh(a.zh)) {
      issues.push(`[${i}] zh mismatch:\n      derived: ${d.zh}\n      actual:  ${a.zh}`);
    }
    if (d.role !== a.role) {
      issues.push(`[${i}] role mismatch: derived=${d.role} actual=${a.role}`);
    }
  }
  return issues;
}

// ── Step 4. Report ────────────────────────────────────────────────────
const report = { perfect: [], minor: [], blocked: [] };

for (let i = 0; i < Math.max(LESSONS.length, LESSONS_META.length); i++) {
  const lesson = LESSONS[i];
  const meta   = LESSONS_META[i];
  const label  = `Lesson ${i + 1} (LESSONS.id=${lesson && lesson.id} | META.id=${meta && meta.id})`;
  console.log('═══', label, '═══');

  if (!lesson || !meta) {
    console.log('  ⚠️  one side missing — cannot audit\n');
    report.blocked.push({ idx: i, reason: 'one side missing' });
    continue;
  }
  if (!Array.isArray(lesson.sentences) || lesson.sentences.length === 0) {
    console.log('  ⚠️  no sentences[] — cannot derive\n');
    report.blocked.push({ idx: i, reason: 'no sentences' });
    continue;
  }
  if (!Array.isArray(meta.dialogue) || meta.dialogue.length === 0) {
    console.log('  ⚠️  no dialogue[] — nothing to compare\n');
    report.blocked.push({ idx: i, reason: 'no dialogue' });
    continue;
  }

  const derived = deriveDialogue(lesson.sentences);
  const issues  = compareDialogue(derived, meta.dialogue);

  console.log(`  sentences=${lesson.sentences.length}  derived=${derived.length}  actual=${meta.dialogue.length}`);
  if (issues.length === 0) {
    console.log('  ✅ perfect match — safe to derive\n');
    report.perfect.push(i);
  } else if (issues.length <= 2 && derived.length === meta.dialogue.length) {
    console.log('  🟡 minor diffs:');
    issues.forEach(s => console.log('    ' + s));
    console.log('');
    report.minor.push({ idx: i, issues });
  } else {
    console.log('  ❌ structural mismatch:');
    issues.slice(0, 8).forEach(s => console.log('    ' + s));
    if (issues.length > 8) console.log('    ...(' + (issues.length - 8) + ' more)');
    console.log('');
    report.blocked.push({ idx: i, reason: 'structural', count: issues.length });
  }
}

// ── Summary ────────────────────────────────────────────────────────────
console.log('\n══════════════ SUMMARY ══════════════');
console.log(`✅ Perfect (safe to derive): ${report.perfect.length}  →  lessons ${report.perfect.map(i => i + 1).join(', ') || '(none)'}`);
console.log(`🟡 Minor diffs (fixable):    ${report.minor.length}    →  lessons ${report.minor.map(o => o.idx + 1).join(', ') || '(none)'}`);
console.log(`❌ Blocked (keep manual):    ${report.blocked.length}    →  lessons ${report.blocked.map(o => o.idx + 1).join(', ') || '(none)'}`);
console.log('\nVerdict:');
if (report.blocked.length === 0 && report.minor.length === 0) {
  console.log('  🟢 ALL lessons can be derived. Safe to migrate to single source.');
} else if (report.blocked.length === 0) {
  console.log('  🟡 Most lessons OK; minor diffs can be normalized before migration.');
} else {
  console.log('  🔴 Some lessons need data fixes OR must keep dual structure for now.');
}
